### ISSUE TITLE
Bitbucket support

### ISSUE THREAD

**Issue by flowdee** [YahnisElsts/plugin-update-checker#48]
Hey @YahnisElsts, any plans to support Bitbucket (private repos) as well? Would be amazing!

**Comment by YahnisElsts** [comment]
There are no such plans at the moment. Not that there's anything wrong with the idea; I just don't have the time.

**Comment by flowdee** [comment]
thanks for the reply anyway!

**Comment by bilaltas** [comment]
Any plan to develop that?

**Comment by dpolyakov** [comment]
+1

**Comment by bilaltas** [comment]
+1

**Comment by YahnisElsts** [comment]
I might have a chance to add this sometime soon, so I'm looking for input. What's your typical workflow for releasing new plugin/theme versions on BitBucket? How would one go about finding the latest stable version? (Is it a branch, tag, or does BitBucket have something like GitHub's "releases"?)

**Comment by bilaltas** [comment]
Great, many thanks in advance! Tags are used for that on Bitbucket.

**Comment by dpolyakov** [comment]
Usually i am adding new tag, like `v2.1`, and merging feature branch in the master branch.

**Comment by bilaltas** [comment]
But you don't have to deal with tags. You just get "readme.txt" from master and parse it.
That's it.

**Comment by YahnisElsts** [comment]
Thanks for the feedback. However, wouldn't you still have to deal with tags to generate download URLs and get the last update timestamp? Also, I suspect that some developers might leave out the "Stable tag" header from their readme.txt.

On an unrelated note, how about authentication? Do you use public repositories (so no auth necessary), or private repositories with username/password, app passwords, or something else?

**Comment by dpolyakov** [comment]
I use private repos

**Comment by bilaltas** [comment]
No, you don't have to. The download url is always same as https://bitbucket.org/{your-user-name}/repo/get/master.zip. To get the last update timestamp, you use the last commit created time.

I'm using free private repositories on Bitbucket. (Most of the people uses it because of that.) So, you need to access BitBucket API with a key and secret which is generated on this link: https://bitbucket.org/account/user/{your-user-name}/api
But, I'm not sure how to get downloadable zip url with that key and secret for Wordpress.

These may help you: 
https://confluence.atlassian.com/bitbucket/oauth-on-bitbucket-cloud-238027431.html
https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/downloads#downloadsResource-GETaspecificfile

**Comment by YahnisElsts** [comment]
@bilaltas Wait, so are you using tags or not? That download URL would point to the master branch, not a tag. Or are you implying that in your case master is always stable and identical to the latest tag?

**Comment by bilaltas** [comment]
@YahnisElsts I'm using Bitbucket Tags but I don't have to. For me, it's just to show which commit is belong to which version of the plugin. Not everyone fills it, most of them do. Yes, in my case, master is always stable and identical to the latest tag. If I work on a beta version, I would create another branch.

https://bitbucket.org/{your-username}/{your-repo}/get/{your-latest-tag}.zip (Tags might be text or number or both)
and 
https://bitbucket.org/{your-username}/{your-repo}/get/master.zip downloads the same.

**Comment by korndev** [comment]
We're using Bitbucket and don't use tags much right now and all our repos are private. The master branch is always production ready. That said, I like to sometimes use a "beta" branch for beta releases on plugins for those that want bleeding edge. If you went the direction of tags though, it'd be more than welcomed in order to integrated with Bitbucket repos.

**Comment by YahnisElsts** [comment]
I think the update checker could support a couple of different approaches like it already does for plugins hosted on GitHub. For example, it could do this (in priority order):

1. Look at `readme.txt` in a specific branch. If it has a `Stable tag` header that points to a valid tag, use that tag.
2. Look at the most recent tags, sort the ones that look like version numbers in descending order, and use the first tag.
3.  Just use the specified branch. The branch is configurable and defaults to `master`.

**Comment by dpolyakov** [comment]
Sounds like a plan :)

**Comment by bilaltas** [comment]
About one year ago, I tried integrating Bitbucket api for your script but I haven't had time to complete. Last status of it has ability to check and detect the new version, but it cannot download and install it yet. I used [Gentlero's Bitbucket PHP API Wrapper](http://gentlero.bitbucket.org/bitbucket-api/). I hope this files help you to complete it: (I'm not sure it still works, but it gives an idea.)
- bitbucket-checker.php
- plugin-update-checker.php

[plugin-update-checker-with-bitbucket.zip](https://github.com/YahnisElsts/plugin-update-checker/files/669251/plugin-update-checker-with-bitbucket.zip)

**Comment by YahnisElsts** [comment]
I actually just got my own implementation to a sort-of-usable state. Anyone interested in testing BitBucket integration can find it in this branch:
https://github.com/YahnisElsts/plugin-update-checker/tree/theme-support

Usage: 
```php
$bitbucketPluginChecker = Puc_v4_Factory::buildUpdateChecker(
	'https://bitbucket.org/username/repository-name',
	__FILE__, //Full path to the main plugin file.
	'unique-plugin-slug' //Usually the same as your plugin directory name.
);

//If your repository is private, create an OAuth consumer that has read access
//to the repository and and set the credentials like this:
$bitbucketPluginChecker->setAuthentication(array(
	'consumer_key' => 'abc',
	'consumer_secret' => 'def',
));
```

Fair warning: This stuff is mostly untested and definitely not production-ready. It might just crash right away.

**Comment by bilaltas** [comment]
Oh good! I didn't know you already started it. :)

**Comment by YahnisElsts** [comment]
Update: Now BitBucket support should also work with themes. The setup is almost identical to plugins:

```php
$bitbucketThemeChecker = Puc_v4_Factory::buildUpdateChecker(
	'https://bitbucket.org/username/repository-name',
	__FILE__, //Full path to your functions.php file.
	'unique-theme-slug' //Optional. Usually the same as your theme directory name.
);
```

The only significant difference is that the second parameter must be an absolute path to any file inside your theme's directory. If your update-related code is inside functions.php, the `__FILE__` constant would be a good choice. The update checker uses this path to distinguish between plugins and themes and to get the theme slug.

**Comment by YahnisElsts** [comment]
Update: Version 4.0 has been released and it includes BitBucket support.

**Comment by bilaltas** [comment]
Thank you Yahnis. But I couldn't get it worked yet. Two problems I saw:

1. Doesn't work with 3 digits version numbers
<img width="72" alt="screen shot 2017-01-11 at 22 40 52" src="https://cloud.githubusercontent.com/assets/12401765/21863578/b639bf58-d846-11e6-9a00-94ad6c45c4a8.png">
Shows v1.0 as the final version.

2. If I change the installed plugin's version to 0.0, it shows an update. Once I start the update, it gives this error message:
<img width="1231" alt="screen shot 2017-01-11 at 22 44 12" src="https://cloud.githubusercontent.com/assets/12401765/21863701/269c4acc-d847-11e6-902b-bf0773211e5d.png">

**Comment by YahnisElsts** [comment]
Thank you for the report. I could swear I tested the installation process multiple times...

Anyway, both of those issues should be fixed now.

**Comment by bilaltas** [comment]
Hello Yahnis! Is there any update on these issues?

**Comment by YahnisElsts** [comment]
As I said, the issues should be fixed now.

**Comment by bilaltas** [comment]
Sorry, I misunderstood you. I think because it was so fast. :)
I tried once you replied, both wasn't working. Now, issue #1 is still exist. It still shows the v0.1 as the latest. And, when I update it, it updates to the v0.1 perfectly. So, issue #2 is fixed.

**Comment by YahnisElsts** [comment]
Well, I can't reproduce the problem with version numbers. In my own tests, the update checker picks the correct version.

Are you sure that:
1. You've got the latest changes from the master branch?
2. The style.css in the v0.1.6 tag has the correct header, i.e. `Version: 0.1.6`?

**Comment by bilaltas** [comment]
1. Yes, I have only branch which is master one.
2. Yes, it has correct version header: Version: 0.1.6 (without "v" letter)

Could it be about the "v0.1" version is two-digit or something like that?

**Comment by YahnisElsts** [comment]
That shouldn't matter. I've tested it with both two-digit and three-digit version numbers, with and without the `v` prefix.

Unfortunately, there's not much else I can do without seeing what exactly is going on in your setup. What I would suggest is this: take a look at the `getLatestTag` method in `Puc/v4/Vcs/BitBucketApi.php`. Try tracing it with a debugger or add some `var_dump()`'s statements to see what it's doing. Confirm that:

1. `$tags->values` includes the `v0.1.6` tag
2. This tag is also the first entry in `$versionTags`. 

Chances are, one of those will be wrong. Then we can move on from there.

**Comment by bilaltas** [comment]
Interestingly, it shows all the versions up to v0.1, not more.

**Comment by bilaltas** [comment]
Only first 10 versions.

**Comment by bilaltas** [comment]
I think I got the problem. The results are paginated as 10 by 10. Once I print the $tags directly, it shows:
<img width="304" alt="screen shot 2017-01-24 at 01 47 54" src="https://cloud.githubusercontent.com/assets/12401765/22226332/ce847daa-e1ce-11e6-9d80-a1d57cc675f2.png">

**Comment by bilaltas** [comment]
Hello again,
I could not find any way to reverse the output on the api. So I tried detecting the last page and updated the api code. I've changed the `getLatestTag` method like this and the problem is solved:
```
public function getLatestTag() {
	$tags = $this->api('/refs/tags');
	if ( !isset($tags, $tags->values) || !is_array($tags->values) ) {
		return null;
	}

	$last_page = floor($tags->size / $tags->pagelen) + 1;
	if ( $last_page > 1 ) {
		$tags = $this->api('/refs/tags?page='.$last_page);
	}

	//Filter and sort the list of tags.
	$versionTags = $this->sortTagsByVersion($tags->values);

	//Return the first result.
	if ( !empty($versionTags) ) {
		$tag = $versionTags[0];
		return new Puc_v4_Vcs_Reference(array(
			'name' => $tag->name,
			'version' => ltrim($tag->name, 'v'),
			'updated' => $tag->target->date,
			'downloadUrl' => $this->getDownloadUrl($tag->name),
		));
	}
	return null;
}
```

**Comment by YahnisElsts** [comment]
It looks like you're right about the source of the bug. I had assumed BitBucket would return the most recent results first, but that's apparently not the case. I didn't catch this in my own testing because my repository didn't have enough tags to trigger pagination.

There is actually a way to [sort query responses](https://developer.atlassian.com/bitbucket/api/2/reference/meta/filtering#query-sort). The documentation is pretty short on details, but after some expermentation, I was able to find a way to sort tags by commit date. (`sort=-target.date`). Try the patch I just committed - it should fix the problem.
